[1단계 - 행운의 로또미션] 티케(조은현) 미션 제출합니다.

devhyun637 opened the pull request:
안녕하세요, Jbee! 1단계 리뷰요청이 많이 늦어져서 죄송합니다. 
콜린(@2SOOY)과 함께 페어를 진행한 티케입니다.
저번에도 `🎰 행운의 로또`를 리뷰 해주셨는데 이번에도 또 로또로 뵙게 되었네요!
~~혹쉬 Jbee가 저의 행운의 로또👀?~~(죄송합니다)

## 📃 [데모페이지](https://2sooy.github.io/react-lotto/)
### 데모 영상
https://user-images.githubusercontent.com/59258239/115600634-e1097200-a317-11eb-8c3d-919d248d1da3.mov


## 🖐 이런 부분에 집중했습니다.
저랑 페어 모두 React가 처음이라서 `왜 리액트라는 것을 사용하는가?`에 대해서 고민을 하면서 그 목적에 따라 코드를 작성하고 싶었습니다.
또 `React를 체험해보`는데 대부분의 시간투자를 한 것 같습니다. 이에 리뷰어님께서 `중점적으로 봐주셨으면 하는 부분`에 대한 감이 안와서, 저희가 집중한 부분에 대해서 정리해보았습니다.

1. UX 증진을 위한 고민
- 처음 프로그램을 시작하면, 금액을 입력하는 input에 `AutoFocus`가 됩니다.
- 당첨 '번호'를 보는 부분에서 많은 양을 구매했을 때, 스크롤이 가능하도록 길이를 제한하였습니다.
- 당첨번호 입력 input type을 `checkbox` 로 만들어, 선택한 숫자를 사용자가 직관적으로 볼 수 있게 하였습니다.
- `다시 시작하기` 버튼을 선택했을 때, `confirm`을 통해 사용자에게 확인하는 과정을 거쳤습니다.
2. 웹 표준과 웹 접근성
 - javascript-lotto때와 같이 웹 접근성을 준수하려고 노력하였습니다. (aria-*사용, Head Tag의 단계 준수 등)
- 로또 금액의 단위에 대한 제한을 위해 `input` 태그의 `step` attribute을 적용하였습니다.
3. React 체험하기
- `Reset Button`을 눌렀을 때 `State`를 초기화 한 이후에 `PriceInputForm` 컴포넌트의 `AutoFoucs`의 기능을 사용하기 위해  `Ref` 속성을사용했습니다. 
- 다른 곳에서도 쓰일 수 있는 `modal`이나, 필요한 속성을 props로 받아서 재활용을 할 수 있는 `button`을 utils component로 분리하였습니다.
### ❓❓❓해결하지 못한 부분❓❓❓
- 현재 `styled component`를 사용하고 있는데, 하나의 컴포넌트 파일에 너무 많은 양의 코드가 들어가는 것 같아서 `[컴포넌트].Style.js`의 형식으로 파일을 하나 더 추가했습니다. 하지만 그 결과, 파일의 수가 늘어나는 것이 고민입니다. 또한, `styled component`를 통해 만든 `element`하고, `컴포넌트`의 구별이 잘 가지 않아서 `styled component`는 이름 뒤에 태그명을 명시하였습니다. (ex. `const AbcDiv = styled.div``;`) 리뷰어님께서 파일 분리한 것과 태그명에 대한 피드백을 주시면 보완하겠습니다.
- React를 React 답게 쓰기 위해서는 `재사용성`이 있는 컴포넌트로 분리해야 된다는 생각이 들었습니다. 이에 한 화면의 컴포넌트를 분리할 수 없을 때까지 분리를 해보고, 굳이 분리를 할 필요가 없는 부분을 다시 합치는 `bottom-up` 방식으로 코드를 구현하였습니다. 하지만 `모든 컴포넌트를 재사용있게 만들어야 하는가?`에 대한 의문을 아직 해결하지 못했습니다. 개인적으로는 모든 컴포넌트를 재사용성을 가지게 하는 것은 도메인에서 분리할 수 없는 부분도 있기 때문에 불가능하다고 생각하는데, 혹시 제가 참고할만한 인사이트가 있다면 알려주시면 감사하겠습니다.

## 📝 구현 목록
### ✅ 미션 필수 요구 사항
- [x] Class Component를 사용합니다.
- [x] 로또 구입 금액을 입력하면, 금액에 해당하는 로또를 발급해야 한다.
- [x] 로또 1장의 가격은 1,000원이다.
- [x] 소비자는 자동 구매만 할 수 있다.
- [x] 로또 번호는 번호보기 토글 버튼을 클릭하면, 볼 수 있어야 한다.
- [x] 결과 확인하기 버튼을 누르면 당첨 통계, 수익률을 모달로 확인할 수 있다.
- [x] 로또 당첨 금액은 고정되어 있는 것으로 가정한다.
- [x] 다시 시작하기 버튼을 누르면 초기화 되서 다시 구매를 시작할 수 있다.

### ✅ 구현 목록
1. 사용자는 금액 입력을 할 수 있다.
- [x] autofocus를 통해 로또 구입 금액 입력에 대한 인지를 할 수 있다.
- [x] 금액은 1,000원 단위만 입력할 수 있다.
- [x] 최소 금액은 1,000원이고 최대 금액은 50,000원이다.

2. 사용자는 로또를 구입하고, 로또 번호를 확인할 수 있다.
- [x] 로또의 번호는 1 ~ 45 사이의 중복되지 않는 번호이다.
- [x] 지불한 금액에 해당하는 로또를 발급받는다.
- [x] 로또 개수 만큼의 티켓 아이콘을 보여준다.
- [x] 번호 보기 버튼을 클릭하면, 구입한 로또 번호를 보여준다.

3. 사용자는 로또 당첨 번호를 입력할 수 있다.
- [x] 로또 당첨 번호는 1 ~ 45 사이의 중복되지 않는 번호이다.
  - [x] 보너스 번호에 중복된 숫자를 입력했을 때, alert로 알린다.
- [x] 로또 당첨 번호는 6개의 당첨 번호와 1개의 보너스 번호로 이루어져있다.

4. 사용자는 당첨 결과를 확인하고, 프로그램을 재시작 할 수 있다.
- [x] 결과 확인 버튼을 선택하면, 모달 창이 열린다.
- [x] 등수 별 당첨 금액과 수익률을 확인할 수 있다.
- [x] ❌ 버튼이나 dimmed 된 영역을 선택하면, 모달 창을 닫을 수 있다.
- [x] 다시 시작하기 버튼을 누르면, 프로그램을 다시 시작한다.

### 🗂 파일 구조
```
📦src
 ┣ 📂components
 ┃ ┣ 📂LottoPurchaseForm
 ┃ ┃ ┣ 📜LottoPurchaseForm.js
 ┃ ┃ ┗ 📜LottoPurchaseForm.style.js
 ┃ ┣ 📂lottoPurchaseResult
 ┃ ┃ ┣ 📜LottoDetailList.js
 ┃ ┃ ┣ 📜LottoDetailList.style.js
 ┃ ┃ ┣ 📜LottoIconList.js
 ┃ ┃ ┣ 📜LottoIconList.style.js
 ┃ ┃ ┣ 📜LottoItem.js
 ┃ ┃ ┣ 📜LottoItem.style.js
 ┃ ┃ ┣ 📜PurchaseResult.js
 ┃ ┃ ┣ 📜PurchaseResult.style.js
 ┃ ┃ ┣ 📜ToggleButton.js
 ┃ ┃ ┗ 📜ToggleButton.style.js
 ┃ ┣ 📂lottoRewardResult
 ┃ ┃ ┣ 📜RestartButton.js
 ┃ ┃ ┣ 📜RestartButton.style.js
 ┃ ┃ ┣ 📜RewardModalInner.js
 ┃ ┃ ┣ 📜RewardResultTable.js
 ┃ ┃ ┗ 📜RewardResultTable.style.js
 ┃ ┣ 📂lottoWinningNumber
 ┃ ┃ ┣ 📜BonusNumberInput.js
 ┃ ┃ ┣ 📜BonusNumberInput.style.js
 ┃ ┃ ┣ 📜NumberList.js
 ┃ ┃ ┣ 📜NumberList.style.js
 ┃ ┃ ┣ 📜WinningNumberForm.js
 ┃ ┃ ┗ 📜WinningNumberForm.style.js
 ┃ ┣ 📂utils
 ┃ ┃ ┣ 📜Button.js
 ┃ ┃ ┣ 📜Modal.style.js
 ┃ ┃ ┗ 📜Modals.js
 ┃ ┣ 📜App.js
 ┃ ┗ 📜App.style.js
 ┣ 📂constants
 ┃ ┣ 📜cssAttribute.js
 ┃ ┣ 📜lotto.js
 ┃ ┗ 📜messages.js
 ┣ 📂services
 ┃ ┣ 📜lottoPurchase.js
 ┃ ┗ 📜winningResult.js
 ┣ 📂utils
 ┃ ┗ 📜index.js
 ┣ 📜index.css
 ┣ 📜index.js
 ┗ 📜reportWebVitals.js
```

2SOOY committed:
Initialize project using Create React App

2SOOY committed:
docs: README 업데이트

2SOOY committed:
chore: 프로젝트 초기 설정

- prettier 적용
- eslint 설치

2SOOY committed:
feat: 어플리케이션 구조 컴포넌트 파일로 분리

devhyun637 committed:
docs: 금액 입력을 위한 기능 목록 업데이트

devhyun637 committed:
feat: 지불 금액 입력 기능 구현

- 안쓰이는 컴포넌트 파일 삭제

2SOOY committed:
feat: LottoPurchaseResult 마크업 구성

Co-authored-by: EUNHYUN CHO <devhyun637@users.noreply.github.com>

2SOOY committed:
refactor: lable 태그 for 속성 htmlFor로 수정

2SOOY committed:
feat: 로또 발급 기능 구현

- price state 제거
- 랜덤 숫자 생성 유틸 함수 추가

2SOOY committed:
docs: 로또 구매 및 번호 확인을 위한 기능 목록 업데이트

devhyun637 committed:
feat: 발급한 로또의 개수만큼의 아이콘을 보여주는 기능 구현

- 로또 발급 여부에 따른 조건부 렌더링 구현

Co-authored-by: Lee Soo Yeon <2SOOY@users.noreply.github.com>

devhyun637 committed:
feat: 번호 보기 클릭에 따른 로또번호를 보여주는 기능 구현

- 로또를 생성할때 'Set -> 배열'

Co-authored-by: Lee Soo Yeon <2SOOY@users.noreply.github.com>

2SOOY committed:
feat: LottoWinningNumber 마크업 구성 및 스타일 적용

Co-authored-by: EUNHYUN CHO <devhyun637@users.noreply.github.com>

2SOOY committed:
style: 섹션간 마진값 수정, <ul>의 마진 제거

2SOOY committed:
docs: 로또 당첨 번호 입력을 위한 기능 목록 업데이트

2SOOY committed:
feat: WinningNumberForm 컴포넌트 세부 컴포넌트로 분리

Co-authored-by: EUNHYUN CHO <devhyun637@users.noreply.github.com>

2SOOY committed:
docs: 전체적인 기능 구현 목록 업데이트

2SOOY committed:
feat: LottoPurchaseForm <input>의 foucs기능 및 step 속성 추가

- resetLottoPurchaseForm 메서드 구현
- step 속성을 이용한 1000원 단위의 금액 유효성 검사

2SOOY committed:
refactor: App.js components 폴더로 이동

Co-authored-by: EUNHYUN CHO <devhyun637@users.noreply.github.com>

2SOOY committed:
feat: LottoWinningNumber 컴포넌트 마크업 및 스타일 수정

2SOOY committed:
feat: 로또 당첨번호 입력 기능 구현

Co-authored-by: EUNHYUN CHO <devhyun637@users.noreply.github.com>

2SOOY committed:
style: 로또 구매 버튼 색상 변경

Co-authored-by: EUNHYUN CHO <devhyun637@users.noreply.github.com>

2SOOY committed:
style: 토글 버튼 스타일 추가 및 로또 상세 보기 스크롤 적용

Co-authored-by: EUNHYUN CHO <devhyun637@users.noreply.github.com>

2SOOY committed:
feat: 당첨 결과 계산 비즈니스 로직 구현

Co-authored-by: EUNHYUN CHO <devhyun637@users.noreply.github.com>

2SOOY committed:
feat: 로또 당첨 결과를 모달로 보여주는 기능 구현

- App의 isModalOpened 상태 추가
- 모달 여닫기 기능 구현
- 프로그램 재시작 기능 구현

2SOOY committed:
docs: 기능 구현 목록 업데이트

2SOOY committed:
style: 구입한 로또를 보여주는 section의 마진 축소

2SOOY committed:
refactor: 불필요한 코드 삭제

devhyun637 committed:
refactor: Modal 컴포넌트화 및 styled Component 파일 분리

- App.style.js 생성 및 기존의 styled Component 분리
- Modal.sytle.js 생성 및 기존의 styled Component 분리

Co-authored-by: Lee Soo Yeon <2SOOY@users.noreply.github.com>

devhyun637 committed:
refactor: Button 컴포넌트화

devhyun637 committed:
refactor: styled Component 파일 분리

- Button 컴포넌트 적용
- RewardResultTable 메서드 분리
- 필요없는 컴포넌트 삭제 및 부모 컴포넌트와 병합

Co-authored-by: Lee Soo Yeon <2SOOY@users.noreply.github.com>

devhyun637 committed:
refactor: html title 변경

devhyun637 committed:
refactor: 상수화 처리 및 적용

Co-authored-by: Lee Soo Yeon <2SOOY@users.noreply.github.com>

2SOOY committed:
feat: App 컴포넌트의 로또 구매 로직 service 레이어로 분리 및 적용

devhyun637 committed:
chore: 배포를 위한 gh-pages 설치 및 안쓰이는 파일 삭제

devhyun637 committed:
refactor: 모든 파일에 prettier 적용

jbee37142 commented on src/services/winningResult.js:
```diff
@@ -0,0 +1,97 @@
+  const totalProfit = counts
+    .filter(count => count !== 0)
+    .reduce((sum, count) => {
+      sum += getProfit(getRank(count));
+      return sum;
```
```suggestion
      return sum + getProfit(getRank(count));
```

요렇게만 해도 충분하지 않나요?

devhyun637 replied:
`reduce` 는 언제써도 계속 어렵네요😢 앞으로는 한번 더 들여다보는 습관을 기르겠습니다. 감사합니다🙇‍♀️

적용커밋 : https://github.com/woowacourse/react-lotto/pull/43/commits/58d9cd6e700d7370ebe2cafd7db606eed45bb087

jbee37142 replied:
사실 reduce가 항상 좋은 것은 아닙니다 ㅎㅎ

jbee37142 commented on src/components/utils/Modals.js:
```diff
@@ -0,0 +1,43 @@
+
+import { ModalSection, ModalInnerDiv, ModalCloseButton } from './Modal.style';
+
+class Modal extends Component {
```
👏 👏 👏 👏 👏 👏 👏 👏 
Modal이 가지고 있어야 하는 관심사만 잘 분리하신 것 같네요 ㅎㅎ

한발 더 나아가서, 이 Modal을 좀 더 말랑말랑하게 만들어볼 수 있을 것 같아요.

`ModalSection`의 props를 모두 받을 수 있도록 해서 (가령 classname) 외부에서 스타일을 조정할 수 있도록 하면 어떨까요?

devhyun637 replied:
👉말랑말랑👈 요새 이 단어에 꽂혔습니다! 
컴포넌트를 말랑말랑하게! 확 와 닿으면서도 막상 구현해보려면 쉽지 않네요 😂
`Modal`안에 있는 `ModalSection`, `ModalInnerDiv`, `ModalCloseButton`의 스타일을 외부에서 조정할 수 있도록 수정했는데, 현재 사용하는 스타일 속성만 `props`로 받고 있습니다. 제가 말랑말랑하게 만든다는 의도대로 제대로 한 게 맞는지, 또 나중에 앱이 커진다면 `어디까지 말랑말랑하게 추상화를 해야하지?`라는 고민이 조금 들었습니다 👀 
(스타일이 추상화를 하려면 margin, padding, border 등 다양해서 모두 props로 받기가 어려울 것 같은데 지금처럼 사용되는 부분에 대해서만 추상화를 해나가면 될까요?)

적용커밋 : https://github.com/woowacourse/react-lotto/pull/43/commits/2e6052d7bb5345aaf9a3f54d3b507c643f7002b3

jbee37142 replied:
> 스타일이 추상화를 하려면 margin, padding, border 등 다양해서 모두 props로 받기가 어려울 것 같은데 지금처럼 사용되는 부분에 대해서만 추상화를 해나가면 될까요?

이 Modal 컴포넌트가 책임져야만 하는 스타일은 어디까지일까? -> 에 대해서만 정의를 하고
그 외 스타일은 전부 override 할 수 있도록 열어보면 어떨까요?

"전부 override 할 수 있도록"은 컴포넌트를 얼마나 열려있게 설계할 것인가에 따라 달라지는 부분이지만 이 Modal은 활짝 열려있어도 괜찮지 않나? 라는 생각에 코멘트 드려봤어요~

jbee37142 commented on src/components/utils/Button.js:
```diff
@@ -0,0 +1,32 @@
+  min-width: ${props => props.size};
+`;
+
+class Button extends Component {
```
이 버튼을 사용할 때, `disabled`처리를 하고 싶어지면 이 props가 또 필요할 것 같은데요 ?ㅎㅎ button의 attributes들을 전부 확장해서 props로 받으면 좀 더 말랑말랑해지지 않을까요?

devhyun637 replied:
추상화(말랑말랑)에 대한 힌트를 많이 주셔서 (`Modal`, `Button`, `Flex`) 이번에 리팩토링하면서 추상화 및 component의 확장을 많이 연습할 수 있었습니다. `styled component`를 사용하니 `style`에 대한 `props`와 `attributes`에 대한 `props`가 많아져서 이렇게까지 해야하나? 싶을 때도 있었지만,  `charkra-ui`도 참고하면서 많이 배웠습니다 :) 

적용커밋 : https://github.com/woowacourse/react-lotto/pull/43/commits/d3ee8d98ff2a4596292c983139c2983853c73f91

jbee37142 commented on src/components/lottoWinningNumber/NumberList.js:
```diff
@@ -0,0 +1,58 @@
+} from './NumberList.style';
+
+class NumberList extends Component {
+  handlePickNumber = e => {
```
핸들러는 정보를 가장 많이 알고 있는 컴포넌트(정보전문가라고 부르기도 해요!)에서 정의되는 것이 좋습니다.
`props.numbers`, `props.setNumbers`와 같이 부모로부터 접근받는 정보를 알아야 한다면 이 핸들러가 이 컴포넌트에 있어야 하는지 의문을 가져볼 수 있습니다.

NumberList는 handleNumberChange라는 props를 전달받아서 pickedNumber를 parameter로 넘겨주는 방식은 어떨까요?

devhyun637 replied:
🔎정보전문가🔍도 정말 좋은 단어 같아요!
그리고 `handler`가 있어야하는 위치에 대해서도 다시 한번 고민을 하게 되었습니다. 전체적으로 `setState`를 넘겨주었던 부분을 `handler`를 넘겨주어서, `parameter`만 받고 `setState`는 `state`를 가지고 있는 곳에서 사용하도록 하였습니다.
~~(아직까지 state를 누가 가지고 있어야하는지, props로 내려주어도 되는 건지 라는 고민을 계속하고 있는건 안비밀..)~~

적용커밋 : https://github.com/woowacourse/react-lotto/pull/43/commits/c6869f8de85cbabcd4983a4f95f0f0d0c540fa78

jbee37142 replied:
> (아직까지 state를 누가 가지고 있어야하는지, props로 내려주어도 되는 건지 라는 고민을 계속하고 있는건 안비밀..)

저 또한 어제도 오늘도 고민하고 있네요 ㅎㅎ 정말 답도 없고 가장 어려운 부분이니 고민하는 습관을 갖는 것이 중요한 것 같습니다~~ 👍

jbee37142 commented on src/components/lottoRewardResult/RestartButton.style.js:
```diff
@@ -0,0 +1,8 @@
+import styled from 'styled-components';
+
+export const RestartButtonWrapperDiv = styled.div`
+  display: flex;
+  justify-content: center;
+  align-items: center;
```
Flex 컴포넌트를 util처럼 만들어두고 확장해서 사용해볼 수도 있을 것 같네요?ㅎㅎ
요런거 참고해보시면 좋은 컴포넌트에 대한 시야가 틔입니다!:https://chakra-ui.com/docs/layout/flex

devhyun637 replied:
Aㅏ, 이거 보면서 엄청 많이 공부했습니다..!🙇‍♀️🙇‍♀️🙇‍♀️🙇‍♀️ Jbee 그저 빛.. 감사합니다❣️
사실 `Modal`과 `Button`를 리팩토링 할때에는 어떻게 `추상화를 해야하지?`,  `확장을 해야하지?`라는 고민이 계속 들었는데, 
보내주신 부분을 참고하면서 구체적으로 감을 잡은게 이 `Flex`입니다.
코드를 처음 작성했을 때도, 반복적으로 사용되지만 조금씩 다른 `flex` layout의 `div`를 `styled component`로 만들면서 비효율적이라는 생각이 많이 들었는데, 이렇게 컴포넌트로 만드니깐 여러 부분을 리팩토링할 수 있었습니다.

현재 `Flex Component`는 `display:flex` 속성을 가지고 있는 `div`에 `flex-flow`, `justify-content`, `align-items`, `flex-wrap`, `flex-basis`, `flex-grow`, `flex-shrink`을 props로 받을 수 있도록 확장했습니다.

적용커밋 : https://github.com/woowacourse/react-lotto/pull/43/commits/eab5ecb8bcf5468211e2eb8b4d550f57c53474cc, https://github.com/woowacourse/react-lotto/pull/43/commits/30071c8e368ac104b513d6111450533343880e10

jbee37142 replied:
👍  가장 실력이 빨리 향상될 수 있는 길은 없지만 오픈소스가 그나마 가장 많은 도움이 됩니다!ㅎㅎ

jbee37142 commented on src/components/utils/Button.js:
```diff
@@ -0,0 +1,32 @@
+  render() {
+    return (
+      <StyledButton
+        type={this.props.buttonType || 'submit'}
```
HTML Button Element와 동일하게 type이라고 하면 어떨까요?
이 컴포넌트를 쓰게 되는 사람은 buttonType은 뭐지? 라고 생각할 수 있을 것 같아요!

devhyun637 replied:
처음에는 구분이 있는 것이 좋지 않을까 (`button`과 `styled Button`과 구분을 짓는 것이 좋지 않을까) 생각했는데, 어차피 쓰는 사람 입장에서는 `button`을 쓰는 것이기 때문에 이름을 통일하였습니다. 감사합니다.🙇‍♀️

적용커밋 : https://github.com/woowacourse/react-lotto/pull/43/commits/21a087a4e429992c9d6e7dc4cb0b87057e4e7e62

jbee37142 commented on src/components/App.style.js:
```diff
@@ -0,0 +1,21 @@
+import styled from 'styled-components';
+
+export const Main = styled.main`
```
FYI) IE엔 main이 없다.
ㅋㅋㅋ

devhyun637 replied:
아..! 시멘틱한 태그를 쓰고 싶었는데, IE 사용자를 고려안한 판단이었네요 😢
이 프로젝트의 가장 중심이 된다는 의미로 `main` 태그를 쓰고 싶었지만 사실상 body 전부를 의미하여, 
`section` 태그만으로도 충분한 것 같아서 수정하였습니다. 

태그 명세를 자세히 보는 습관을 들이겠습니다! (정말 IE 최근 버전도 지원을 안하는지 몰랐습니다...!)

적용커밋 : https://github.com/woowacourse/react-lotto/pull/43/commits/fa2e6d27207772cd187b97b210a34f2f86927dfa

jbee37142 commented on src/components/utils/Button.js:
```diff
@@ -0,0 +1,32 @@
+      <StyledButton
+        type={this.props.buttonType || 'submit'}
+        size={this.props.size}
+        onClick={this.props.clickHandler}
```
clickHandler대신 onClick은 어때요?ㅎㅎ

devhyun637 replied:
[여기 코멘트](https://github.com/woowacourse/react-lotto/pull/43#discussion_r618321115)와 같이 onClick으로 수정했습니다 :)

jbee37142 commented on src/components/lottoRewardResult/RestartButton.js:
```diff
@@ -0,0 +1,28 @@
+
+class RestartButton extends Component {
+  handleRestart = () => {
+    if (window.confirm(MESSAGE.CONFIRM_RESTART)) {
+      this.props.initState();
+      this.props.purchaseForm.resetLottoPurchaseForm();
+    }
```
restart라는 도메인은 부모가 알고 이를 트리거할 handler만 내려받으면 어떨까요?

이 `RestartButton`이란 컴포넌트는 restart가 어떤 과정을 통해 진행되는지(confirm -> initState -> resetForm)를 하고 알 필요가 없다고 생각해요 ㅎㅎ

devhyun637 replied:
오 이런것도 `관심사의 분리`로 볼 수 있을까요? `setState`를 자식이 해주었던 것을 부모가 하게 코드를 리팩토링 했던 것과 마찬가지로(handler를 넘겨주기), 이 부분도 restart의 로직을 `App` 컴포넌트가 가지고 있고(`handleRestart`), 이를 자식 컴포넌트는 props로 내려 받는 형식으로 코드를 리팩토링 하였습니다.

적용커밋 : https://github.com/woowacourse/react-lotto/pull/43/commits/c6869f8de85cbabcd4983a4f95f0f0d0c540fa78

jbee37142 replied:
관심사라는게 귀에 걸면 귀걸이 코에 걸면 코걸이라 '책임'이라는 단어를 사용하곤 합니다!

관심사의 분리 -> 책임의 분리라고 해석될 수도 있고 HTML/CSS/JS 언어에 따른 관심사의 분리 이렇게 이해될 수도 있어서요!

캡슐화? 가 더 어울릴 것 같네요 ㅎㅎ

jbee37142 reviewed (changes requested):
수고 많으셨습니다! 잘하시네요
(중복되는 코멘트는 생략했습니다)

---

PR body에 올려주신 질문에 대한 짧은 답변입니다.

> 리뷰어님께서 파일 분리한 것과 태그명에 대한 피드백을 주시면 보완하겠습니다.

오픈소스들은 어떻게 네이밍을 하고 있는지 어떻게 스타일링 코드를 관리하고 있는지 살펴봐도 좋을 것 같습니다!ㅎㅎ 구글에 github: charkra-ui 한 번 검색해보셔도 좋을 것 같아요! (제가 좋아하는 오픈소스입니다)

> 하지만 모든 컴포넌트를 재사용있게 만들어야 하는가?에 대한 의문을 아직 해결하지 못했습니다

너무 중요하고 어려운 질문이네요 ㅎㅎ

1. 컴포넌트들을 '합성'가능하도록 만드는 것에 집중해보세요!
2. 저는 개인적으로 좋아하진 않지만 Atomic Design이라는 것에 대해서 살펴보면 좋을 것 같아요. (적용보단 공부를 위한 목적)
3. 도메인 간의 응집도를 높이고 결합도를 낮출 수 있도록 끊임없이 고민해보세요! <- 저도 맨날 하는 중!
4. 컴포넌트를 추상화해보세요! <- 이 말은 정말 어려운 말입니다 ㅠㅠ

devhyun637 committed:
refactor: reduce 함수 코드 축약

devhyun637 committed:
refactor: 컴포넌트 속성명 변경

- clickHandler -> onClick
- buttonType -> type

devhyun637 committed:
refactor: Flex 속성을 가지고 있는 컴포넌트 생성 및 적용

devhyun637 committed:
refactor: handler 위치 변경 및 props 구조 분해 할당

- state의 정보를 가장 많이 알고있는 컴포넌트로 handler 위치 변경
- render 안에서 사용되는 props, state 구조 분해 할당

devhyun637 committed:
refactor: Main 태그 Section 태그로 변경

devhyun637 committed:
refactor: styles 폴더 생성 및  styled component 파일 분류

devhyun637 committed:
refactor: Modal 스타일을 외부에서 조정 가능하게 수정

devhyun637 committed:
refactor: Button component 확장

devhyun637 committed:
refactor: 모달 닫기 버튼 웹 접근성을 가지게 버튼 이름 추가

devhyun637 committed:
refactor: Flex component 확장 및 기존의 flex layout의 div에 적용

devhyun637 committed:
refactor: state 이름 구체화 (isToggled -> isShowNumbers)

devhyun637 committed:
refactor: 필요없는 객체 리터럴 삭제 및 css 상수 삭제

devhyun637 commented:
안녕하세요, Jbee님! 
리뷰반영을 하면서, 누가 `state`를 가져야하는지 또 `setState`와 같은 `state`를 관리하는 로직이 어떤 컴포넌트에 있어야 하는지 많이 고민을 하면서 리팩토링을 해나갔습니다.🔨 

부모(App)이 `state`를 관리하고 있어 로직을 부모 컴포넌트에 두었는데, 아직 감이 확실하게 오지 않아서 2단계때 함수형으로 만들어보면서 더 보완해나가겠습니다.

### 리팩토링 한 부분
1. reduce 함수 필요없는 코드삭제 : https://github.com/woowacourse/react-lotto/pull/43/commits/58d9cd6e700d7370ebe2cafd7db606eed45bb087
2. 컴포넌트 속성 명을 일관성 있게 변경 : https://github.com/woowacourse/react-lotto/pull/43/commits/21a087a4e429992c9d6e7dc4cb0b87057e4e7e62
3. flex layout div 컴포넌트 생성  :  https://github.com/woowacourse/react-lotto/pull/43/commits/eab5ecb8bcf5468211e2eb8b4d550f57c53474cc, https://github.com/woowacourse/react-lotto/pull/43/commits/30071c8e368ac104b513d6111450533343880e10
4. handler 로직 위치 변경 : https://github.com/woowacourse/react-lotto/pull/43/commits/c6869f8de85cbabcd4983a4f95f0f0d0c540fa78 
5. 웹 표준에 지키기(Main 태그 -> Section 태그) : https://github.com/woowacourse/react-lotto/pull/43/commits/fa2e6d27207772cd187b97b210a34f2f86927dfa
6. 웹 접근성 기능 추가(아이콘만 있는 버튼에 보이지 않는 이름 부여) : https://github.com/woowacourse/react-lotto/pull/43/commits/e94134a57835ce4961d5d2aa1ef39366606909a4 
참고한사이트 : [MDN button](https://developer.mozilla.org/ko/docs/Web/HTML/Element/button#%EC%A0%91%EA%B7%BC%EC%84%B1_%EA%B3%A0%EB%A0%A4%EC%82%AC%ED%95%AD)
7. 컴포넌트의 재사용성을 높이기 위한 확장 (컴포넌트 말랑말랑하게 만들기) : https://github.com/woowacourse/react-lotto/pull/43/commits/2e6052d7bb5345aaf9a3f54d3b507c643f7002b3, https://github.com/woowacourse/react-lotto/pull/43/commits/d3ee8d98ff2a4596292c983139c2983853c73f91
 
### 공부하면서 궁금했던 점
지금 `payment` 미션을 함께 진행하면서 `storybook` 을 활용하고 있는데, `storybook`이 Jbee가 언급해주신 `Atomic Design`이 아닐까 하는 생각이 듭니다. 사실 react라고 하면 `재사용성`, `component driven development`이 가장 큰 장점이라고 생각했는데, Jbee는 어떤 부분에서 `Atomic Design`을 `별로 안좋아하시는지` 궁금하네용 👀 (실제 개발은 어떻게 하시는지도 궁금합니다..! ~~우테코 특강 어떠신가요~~)

jbee37142 commented:
> 지금 payment 미션을 함께 진행하면서 storybook 을 활용하고 있는데, storybook이 Jbee가 언급해주신 Atomic Design이 아닐까 하는 생각이 듭니다. 사실 react라고 하면 재사용성, component driven development이 가장 큰 장점이라고 생각했는데, Jbee는 어떤 부분에서 Atomic Design을 별로 안좋아하시는지 궁금하네용 👀 (실제 개발은 어떻게 하시는지도 궁금합니다..! 우테코 특강 어떠신가요)

질문의 수준이 남다르네용.

> react라고 하면 재사용성, component driven development이 가장 큰 장점

Vanillajs(또는 jQuery)로도 CDD는 충분히 할 수 있습니다! 당연히 재사용 가능한 컴포넌트를 만드는 것도 포함입니다.
React는 이를 선언적으로 프로그래밍 할 수 있도록 도와주는 도구일 뿐입니다 ㅎㅎ

> 어떤 부분에서 Atomic Design을 별로 안좋아하시는지 궁금하네용 👀

음 (저도 잘은 모르지만ㅎㅎ) atom은 그렇다 치고 molcules(?), organism (?) 이 두 단계가 의미없다고 생각해서 좋아하지 않습니다.
실제로 프로덕트를 만들다 보면 atom, molcules, organism 이렇게 세 단계로 딱 떨어지지 않습니다.
그런데 이런 상황에서 우리는 atomic design을 적용했고 이걸 지켜야해! 하는 순간 구조가 망가지고 제어가 역전되는 상황이 발생합니다.

인터뷰에서도 atomic design을 적용하신 분이 계셔서 여쭤보면 molcules, organism 이 두 가지에 대한 구분이 사람마다 달라 혼란을 겪었다는 이야기도 많이 듣습니다.

그럼에도 왜 atomic design을 적용하나? 하는 부분이 있는데, 그건 복잡한 컴포넌트를 어떻게 추상화할지 고민할 시간에 정해진 규약에 쑤셔넣자? 정도로 이해하고 있습니다 ㅎㅎ

jbee37142 commented on src/components/utils/Flex.js:
```diff
@@ -0,0 +1,34 @@
+
+class Flex extends Component {
+  render() {
+    const {
+      children,
+      flexFlow,
+      justifyContent,
+      alignItems,
+      flexWrap,
+      flexBasis,
+      flexGrow,
+      flexShrink,
+    } = this.props;
+
+    return (
+      <FlexBox
+        FlexBox={flexFlow}
+        justifyContent={justifyContent}
+        alignItems={alignItems}
+        flexWrap={flexWrap}
+        flexBasis={flexBasis}
+        flexGrow={flexGrow}
+        flexShrink={flexShrink}
+      >
+        {children}
+      </FlexBox>
```
요렇게만 해도 충분해보입니다?
```suggestion
    return (
      <FlexBox {...this.props} />
```

jbee37142 commented on src/components/lottoWinningNumber/NumberList.js:
```diff
@@ -0,0 +1,40 @@
+
+class NumberList extends Component {
+  render() {
+    const { numbers, handleChangeNumbers } = this.props;
```
props에서 handle-*로 정의되는데 내부에서 정의도니 handler로 handle-*로 정의되면 혼란럽지 않을까? 하는 생각이 들어요!
props로 전달하는 naming convention을 on-*으로 맞춰봐도 괜찮겠다 싶어요!

jbee37142 reviewed (approved):
여쭤보신 내용에 대해서 + 리뷰 반영해주신 것에 대해 코멘트 좀 달았어요!
수고하셨고 다음 미션에서는 더 좋은 PR 기대해볼게요!
